lib/fast_jsonapi/serialization_core.rb

```diff
@@ -26,6 +26,21 @@ def ids_hash(ids, record_type)
         id_hash(ids, record_type) # ids variable is just a single id here
       end
 
+      def id_hash_from_record(record)
+        { id: record.id.to_s, type: record.class.name.underscore.to_sym }
```

Comment:
@vovimayhem Thanks for tackling this!  I'm wondering if `record.class.name.underscore.to_sym` is making assumptions about what the type should be: it must be determined by the class name, it must be underscored, etc.  It seems like the canonical value for this record type would be on the serializer itself, e.g. `PersonSerializer#record_type`.  If the serializer does not exist, it could potentially fall back to the underscored class name.

It looks like the primitives for accessing other serializers are built in... something like `compute_serializer_key(record.class.name).constantize`.  It still makes the assumption that the serializer name can be inferred from the record class, but there needs to be _some_ sort of convention for a polymorphic lookup...

Reply:
Hmmm... What I can see is that the most brutal performance gains this gem is by reducing the assumptions/choices to be made at serializing time, and passing it instead to the serializer definition time... hence why we have here options such as `belongs_to :owner, record_type: :user` (here we're assuming the object type is `user`, so we don't waste time figuring out the serializer to use at serializing time).

We can't do that with a non-homogeneous collection (i.e. polymorphic association) because there's no way to know exactly which object types are in a collection at definition time...

...but may I suggest something in the middle: We *may* know some (if not all) of the types we expect to see in a collection. That way, if the type of object in a collection is known, we won't need to figure the record_type/serializer out (that `record.class.name.underscore.to_sym`).

I'll give that a try

Reply:
Awesome, thanks for determining the performance bottleneck.

I offered a suggestion in #49 that I think could apply nicely here as well.  Can we define a class method for determining the record type, given the record instance?  For example:

```ruby
def self.polymorphic_record_type(record)
  ...
end
```

The default implementation can use the same methodology for determining the default record type (and potentially even cache the results as key/value pairs rather than calculating them every time), but it can be overridden for those (rare?) cases where you have a polymorphic association whose type is not the default.  It seems like this solution would provide sensible defaults without sacrificing performance, and also have the ability to easily override when the situation requires it.

Reply:
I'm gonna take a look at this.

Reply:
I'm deffinitely going to take a look at this... thanks for the idea!